rfc/rfc-43/rfc-43.md

```diff
@@ -0,0 +1,257 @@
+
+3. Using independent compaction/clustering job is a better way to schedule the job, in this mode the job execute async
+   and do not sharing resources with writing job, but also has some questions:
+    1. Users have to enable lock service providers so that there is not data loss. Especially when compaction/clustering
+       is getting scheduled, no other writes should proceed concurrently and hence a lock is required.
+    2. The user needs to manually start an async compaction/clustering application, which means that the user needs to
+       maintain two jobs.
+    3. With the increase in the number of HUDI jobs, there is no unified service to manage compaction/clustering jobs (
+       monitor, retry, history, etc...), which will make maintenance costs increase.
+
+With this effort, we want to provide an independent compaction/clustering Service, it will have these abilities:
+
+- Provides a pluggable execution interface that can adapt to multiple execution engines, such as Spark and Flink.
+
+- With the ability to failover, need to be persisted compaction/clustering message.
+
+- Perfect metrics and reuse HoodieMetric expose to the outside.
+
+- Provide automatic failure retry for compaction/clustering job.
+
+## Implementation
+
+![](service.jpg)
+
+### Client
+
+Register to Service when the job starts.
+
+### Request Handler
+
+Receive the client's request and save the register message in storage.
+
+### Storage
+
+#### Lectotype
+
+**Requirements:** support single row ACID transactions. Almost all write operations require it, like operation creation,
+status changing and so on.
+
+There are the candidates,
+
+**Hudi table**
+
+pros:
+
+- No external components are introduced and maintained.
+
+crons:
+
+- Each write to hudi table will be a deltacommit, this will further lower the number of possible requests / sec that can
+  be served.
+
+**RDBMS**
+
+pros:
+
+- database that is suitable for structured data like metadata to store.
+
+- can describe the relation between many kinds of metadata.
+
+crons:
+
+- introduce another system to maintain.
+
+**File system**
+
+pros:
+
+- No external components are introduced and maintained.
+
+crons:
+
+- not suitable for the situation that requires high performance.
+
+- have to do extra work to support the metadata organization.
+
+**Key-value storage**
+
+pros:
+
+- database that is suitable for structured data like metadata to store.
+
+- in-memory data store so that read and write faster.
+
+crons:
+
+- introduce another system to maintain.
+
+- stroage capacity is a limitation.
+
+Through the storage of server is pluggable, considering the general situation of disk storage, good performance of read
+and write, convenience of development, RDBMS may be a better one to be chosen.
+
+### Scheduler
+
+- Periodically scan the storage and submit operation job according to user-specified rules, like priority, queue, owner
```

Comment:
Can we trigger this based on events happening on the table i.e look at what new table services need to be scheduled on every commit?  This is what happens in the write client today, again

Reply:
We can trigger based on the commit event and configure a parameter of the maximum number of threads to control the maximum number of tasks submitted at the same time.